File: 프로그래머스/Level 2/뒤에 있는 큰 수 찾기-review-solve.py
"""
numbers: 정수로 이뤄진 배열
배열의 각원소들에 대해 자신보다 뒤에 있는 숫자중에서 자신보다 크면서 가장 가까이 있는 수를 뒷 큰수
return : 모든 원소에 대한 뒷 큰수들을 차례대로 담은 배열
    - 뒷큰수 존재하지 않는 원소는 -1 담음 
"""

def solution(numbers):
    big_nums = [0] * len(numbers)
    stack = [] 
    stack.append((0, numbers[0]))
    
    for i in range(1, len(numbers)):
        while stack and stack[-1][1] < numbers[i]:
            idx, num = stack.pop()
            big_nums[idx] = numbers[i]
        stack.append((i, numbers[i]))
    
    while stack:
        idx, num = stack.pop()
        big_nums[idx] = -1
    
    return big_nums

File: 프로그래머스/Level 2/뒤에 있는 큰 수 찾기-review-solve.java
import java.util.Stack;

class Solution {
    public int[] solution(int[] numbers) {
        int[] bigNums = new int[numbers.length];
        Stack<int[]> stack = new Stack<>();
        
        for (int i = 0; i < numbers.length; i++) {
            while (!stack.isEmpty() && stack.peek()[1] < numbers[i]) {
                int[] elem = stack.pop();
                bigNums[elem[0]] = numbers[i];
                
            }
            stack.push(new int[] {i, numbers[i]});
        }
        
        while (!stack.isEmpty()) {
            int[] elem = stack.pop();
            bigNums[elem[0]] = -1;
        }
        
        return bigNums;
    }
}